ykcapi/src/sginterp.rs

```diff
@@ -0,0 +1,142 @@
+use llvm_sys::bit_reader::LLVMParseBitcodeInContext2;
+use llvm_sys::core::*;
+use llvm_sys::prelude::{LLVMModuleRef, LLVMValueRef};
+use llvm_sys::{LLVMOpcode};
+use std::collections::HashMap;
+use std::ffi::CStr;
+use std::mem::MaybeUninit;
+
+use crate::llvmapihelper::{self};
+
+/// Stopgap interpreter values.
+#[derive(Debug)]
+pub enum SGValue {
+    U32(u32),
+    U64(u64),
+}
+
+/// A frame holding live variables.
```

Comment:
This is going to be very slow: can't we convert variables to indexes in a `Vec` rather than IDs in a `HashMap`?

Reply:
I don't understand how we could do that.

Given a variable `%myvar`, which is identified by a `Value *`, what would it's index be?

Reply:
Every function has a finite set of variables that we know in advance, so it's clearly doable. [It might be awkward, but that's another matter.]

Reply:
I'm not sure if variable order is stable over serialisation or not.

Since this is essentially a question of optimisation, I suggest doing as Lukas has, and adding an `OPT:` referencing an issue.

Reply:
`OPT` (or is it `YKOPT`?) works for me here.

Reply:
Most values will be instructions, which do have an order (bbidx:instridx), which should be stable over serialisation. I'm not sure about constants though, they could appear in any order, even after serialisation.

Reply:
This is about variables though, not values?

Reply:
Yeah, I guess we could put constants in a separate map and then put variables in a vector. Seems doable, but will probably remain a `YKOPT` for now.

Reply:
I'm confused: surely this `HashMap` is from variable names to values? Constants are just a value, right?

Reply:
This map is from `* Value`s to interpreter value. Constants are also just values, but since I am not sure if their order remains stable over serialisation they probably will need to stay in a hashmap, but we certainly can move non constant variables (i.e. instructions) into a vector to increase performance.

Reply:
OK, then I think this is a design issue that I don't understand: LLVM defines variables (`%x` etc) and surely that's what we should be storing as keys, rather than any possible value? After all `2 = 3` (where `2` and `3` are constants) hardly makes sense.

Reply:
> I guess we could put constants in a separate map

Then there's global variables. It's going to get complicated. I suggest we stick to a unified way of identifying variables.

> surely this HashMap is from variable names to values? 

LLVM doesn't identify variables using a name. It uses an opaque pointer value (which obviously isn't stable across serialisation).

Whilst you can assign variables names, there's no guarantee that the will carry into serialised IR. And if variables have no names and you print the IR, then the printer itself assigns them numeric names on the fly.

Reply:
Global variables shouldn't be stored in the frame, by definition, right?

My claim is that (whether they're in a generic `Value` struct or not), the only valid keys are variable "name"s, not constants, nor globals.

Reply:
> Global variables shouldn't be stored in the frame, by definition, right?

Ah, that's correct, yes.

> My claim is that (whether they're in a generic Value struct or not), the only valid keys are variable "name"s, not constants, nor globals.

The only valid variables in a live frame are local variables, but I don't understand this "name" thing.

Reply:
Oh, yes, you're right, we don't need to store globals and consts in the frame at all. That being said, we still can't use variable "names" as keys, since (as @vext01 pointer out above) there's no guarantee they have one. So it'll have to be pointers for now. But we'll later change the variable map to a vector, then key goes away anyway.

Reply:
It's standard CS theory: variables have names. Names don't have to be strings: they can be numbers, or pointers, or whatever.

Reply:
I've never heard of that, but if we define an LLVM variable name as its opaque pointer, then yes, this map is keyed by "names".

Is the confusion then, that an LLVM "variable name" has a specific meaning in LLVM, that is different from CS theory?

Reply:
I think the confusion is because we assume "name" must be "string", but that needn't be the case: it just needs to be something that uniquely identifies that variable.

The challenge we have here -- and why I would suggest we might want to use a key that *isn't* an LLVM `Value` directly -- is that (in a normal language) we've got the equivalent of `HashMap<Object, Object>` which is why we end up thinking you can have constants or global variables as keys (when you can't). I would suggest having a struct something like `SGVarName` (or maybe even change the serialisation to make this clear!) so that we end up with `HashMap<VarName, Value>` to make clear that only a strict subset of possible `Value`s can be keys.

Reply:
Just to double check: d79ec17 does what you meant?